### PR TITLE
adding root support functionality

### PR DESCRIPTION
**Pull request by aelnaiem**
This code ended up being a mess...

[asana task](https://app.asana.com/0/201972293666219/198715212854040)

### CODE REVIEW

File: lib/evaluateFunctions.js [hunk 1]
```diff
@@ -65,6 +68,229 @@ function evaluateFunctions(node) {
...
+      }
+
+      return NodeStatus.nodeChanged(
+        MathChangeTypes.NTH_ROOT_VALUE, oldNode, newNode, true, subSteps);
+    }
+    else if (radicandNode.op === '^') {
+      const baseNode = radicandNode.args[0];
+      const exponentNode = radicandNode.args[1];
+      // TODO(ael): this only works if they actually have a value...
+      // ideally I'd use equals, but the nodes are different for some reason
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: equals() isn't working? or ===

I know that === doesn't work, but I'm surprised about .equals()

File: lib/evaluateFunctions.js [hunk 2]
```diff
@@ -65,6 +68,229 @@ function evaluateFunctions(node) {
...
+  const radicandNode = newNode.args[0];
+  const rootNode = newNode.args.length === 2 ? newNode.args[1] : NodeCreator.constant(2);
+
+  if (NodeType.isOperator(radicandNode)) {
+    // if is operator and operation is addition, we can't do anything
+    if (radicandNode.op === '+' || radicandNode.op === '-') {
+      return NodeStatus.noChange(node);
+    }
+    else if (radicandNode.op === '*' || radicandNode.op === '/') {
+      // distribute the root into the children
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: I'm not sure why this is a step - some tests with substep examples and examples of math in the code comments would both help

Comment (evykassirer) [inline lib/evaluateFunctions.js]: because `sqrt(2*18)` is only possible if you have the next step `sqrt(2*2*3*3)`

File: lib/evaluateFunctions.js [hunk 3]
```diff
@@ -65,6 +68,229 @@ function evaluateFunctions(node) {
...
+    else if (radicandNode.op === '^') {
+      const baseNode = radicandNode.args[0];
+      const exponentNode = radicandNode.args[1];
+      // TODO(ael): this only works if they actually have a value...
+      // ideally I'd use equals, but the nodes are different for some reason
+      // even if the value is the same...
+      if (rootNode.value === exponentNode.value) {
+        newNode = baseNode;
+        return NodeStatus.nodeChanged(
+          MathChangeTypes.NTH_ROOT_VALUE, oldNode, newNode);
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: I'd make a special change type for this! it's a fun edge case

File: lib/evaluateFunctions.js [hunk 4]
```diff
@@ -65,6 +68,229 @@ function evaluateFunctions(node) {
...
+      } else {
+        return NodeStatus.noChange(node);
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: there are some other cases we'd want to catch to probably

mainly I'm thinking of exponent % rootnode value === 0

e.g. `sqrt(x^4)`

File: lib/evaluateFunctions.js [hunk 5]
```diff
@@ -65,6 +68,229 @@ function evaluateFunctions(node) {
...
+      newNode = NodeCreator.constant(nthRootValue);
+      return NodeStatus.nodeChanged(
+        MathChangeTypes.NTH_ROOT_VALUE, oldNode, newNode);
+    }
+    else {
+      const subSteps = [];
+
+      // 1. convert the number into the product of its prime factors
+      const factors = Factor.getPrimeFactors(radicandValue);
+      const factorNodes = factors.map(factor => {
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: fun javascript neatness: you can do `factors.map(NodeCreator.constant)`

File: lib/evaluateFunctions.js [hunk 6]
```diff
@@ -65,6 +68,229 @@ function evaluateFunctions(node) {
...
+        return NodeCreator.constant(factor);
+      })
+
+      newNode.args[0] = NodeCreator.operator('*', factorNodes);
+      subSteps.push(NodeStatus.nodeChanged(
+          MathChangeTypes.FACTOR_INTO_PRIMES, oldNode, newNode));
+
+      // TODO: Consider moving steps 2, 3 and 4 into distributing the
+      // radical into children above, and just calling nthNode on the
+      // factored version
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: not sure what you're talking about here

File: lib/evaluateFunctions.js [hunk 7]
```diff
@@ -65,6 +68,229 @@ function evaluateFunctions(node) {
...
+          if (factors[i] !== factors[j]) {
+            break;
+          }
+          j++;
+        }
+        if (j - i >= rootValue) {
+          let groupEnd = i + rootValue;
+          groupedFactors.push(factors.slice(i, groupEnd));
+          hasGroups = true;
+          i = groupEnd;
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: since you reset i to groupEnd, you may as well break out of the while loop once j-i == rootValue (that might be off by one but you get the idea)

File: lib/evaluateFunctions.js [hunk 8]
```diff
@@ -65,6 +68,229 @@ function evaluateFunctions(node) {
...
+        }
+        else {
+          groupedFactors = groupedFactors.concat(factors.slice(i, j));
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: consider having a list of factor group of the size we want, and a separate group of remaining factors we can put at the end?
or would you rather keep the factors all in the same order and just insert the parens

File: lib/evaluateFunctions.js [hunk 9]
```diff
@@ -65,6 +68,229 @@ function evaluateFunctions(node) {
...
+      }
+
+      if (!hasGroups) {
+        return NodeStatus.noChange(node);
+      }
+
+      let children = [];
+      for (let i = 0; i < groupedFactors.length; i++) {
+        const item = groupedFactors[i];
+        if (groupedFactors[i].constructor === Array) {
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: can this happen even if there's more than one but less than the radicandValue?

File: lib/evaluateFunctions.js [hunk 10]
```diff
@@ -65,6 +68,229 @@ function evaluateFunctions(node) {
...
+      // 3. convert grouped factors into exponent nodes
+      oldNode = NodeStatus.resetChangeGroups(newNode);
+      newNode = clone(oldNode, false);
+
+      for (let i = 0; i < children.length; i++) {
+        const child = children[i];
+        if (NodeType.isParenthesis(child)) {
+          const childNodeStatus = convertMultiplicationToExponent(child.content);
+          newNode.args[0].args[i] = childNodeStatus.newNode;
+          subSteps.push(NodeStatus.childChanged(newNode, childNodeStatus, i));
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: the childChanged also updates newNode if you want, but then you need to do a newNode = status.newNode after

File: lib/evaluateFunctions.js (hunk 9 above)
Comment (evykassirer) [inline lib/evaluateFunctions.js]: ohh I see later why you do this
might not be obvious to the student at this stage. pedagogy details though, can always change later

File: lib/evaluateFunctions.js [hunk 11]
```diff
@@ -65,6 +68,229 @@ function evaluateFunctions(node) {
...
+        }
+      }
+
+      // 4. move grouped factors outside of the radical
+      oldNode = NodeStatus.resetChangeGroups(newNode);
+      newNode = clone(oldNode, false);
+
+      const newArgs = [];
+      const newRadicandArgs = [];
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: confused about the difference between these two and why you always push to newRadicandArgs

comment examples would defs help

File: lib/evaluateFunctions.js [hunk 12]
```diff
@@ -96,4 +99,389 @@ function absoluteValue(node) {
...
+    if (status.hasChanged()) {
+      substeps.push(status);
+      newNode = NodeStatus.resetChangeGroups(status.newNode);
+    }
+
+    return NodeStatus.nodeChanged(
+      MathChangeTypes.NTH_ROOT_VALUE, node, newNode, true, substeps);
+  }
+
+  return NodeStatus.noChange(node);
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: ohh okay so you put it back together in combineRoots, but that's not always called after the distribute right?

File: lib/evaluateFunctions.js [hunk 13]
```diff
@@ -96,4 +99,389 @@ function absoluteValue(node) {
...
+// e.g. 2 * nthRoot(2) * nthRoot(x) -> 2 * nthRoot(2 * x)
+function combineRoots(node) {
+  let newNode = clone(node, false);
+  const rootNode = node.args.length === 2 ? node.args[1] : NodeCreator.constant(2);
+
+  const children = [];
+  const radicandArgs = [];
+  for (let i = 0; i < newNode.args.length; i++) {
+    const child = newNode.args[i];
+    if (NodeType.isFunction(child, 'nthRoot') && child.args[1].equals(rootNode)) {
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: what it doesn't have args[1] in the other case of  `node.args.length === 2 ? node.args[1] : NodeCreator.constant(2);`

Comment (evykassirer) [inline lib/evaluateFunctions.js]: maybe make a function for getting the rootNode of a function root node

File: lib/evaluateFunctions.js [hunk 14]
```diff
@@ -96,4 +99,389 @@ function absoluteValue(node) {
...
+      MathChangeTypes.EVALUATE_DISTRIBUTED_NTH_ROOT, node, newNode, true, substeps);
+  }
+
+  return NodeStatus.noChange(node);
+}
+
+// Given a multiplication nodes, including nthRoots, will combine the nodes
+// with the same radicand as a step
+// e.g. 2 * nthRoot(2) * nthRoot(x) -> 2 * nthRoot(2 * x)
+function combineRoots(node) {
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: `e.g. 2 * nthRoot(2) * nthRoot(x) ->`

does that mean `node` is a multiplication of root nodes? or a single root node? some of this code seems like it's as if the former is true and some for the latter, though maybe I'm missing something

File: lib/evaluateFunctions.js [hunk 15]
```diff
@@ -96,4 +99,389 @@ function absoluteValue(node) {
...
+// nthRootMultiplication on the new nthRoot
+function nthRootConstant(node) {
+  let newNode = clone(node, false);
+  const radicandNode = node.args[0];
+  const rootNode = node.args.length === 2 ? node.args[1] : NodeCreator.constant(2);
+
+  if (Negative.isNegative(radicandNode)) {
+    return NodeStatus.noChange(node);
+  }
+  else if (!NodeType.isConstant(rootNode) || Negative.isNegative(rootNode)) {
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: sqrt(-36) is still 6 * root(-1), so I think negative should be allowed and the -1 can be part of the prime factoring

also you should probably check integer and not just constant

File: lib/evaluateFunctions.js [hunk 16]
```diff
@@ -96,4 +99,389 @@ function absoluteValue(node) {
   }
 }
 
+// Evaluate nthRoot() function.
+// Returns a NodeStatus object.
+function nthRoot(node) {
```
Comment (aelnaiem) [inline lib/evaluateFunctions.js]: Yea, I discuss re-organizing modules in the mathsteps organization doc, let's discuss that more

File: lib/evaluateFunctions.js (hunk 12 above)
Comment (aelnaiem) [inline lib/evaluateFunctions.js]: This actually works fine because there's only steps if we call combine roots

Comment (evykassirer) [inline lib/evaluateFunctions.js]: ohhh nice :) can you add some tests for this case?

File: test/evaluateFunctions.test.js [hunk 17]
```diff
@@ -14,6 +14,29 @@ function testEvaluateFunctions(exprStr, outputStr) {
       outputStr);
   });
 }
+describe('nthRoot', function () {
+  const tests = [
+    ['nthRoot(4)', '2'],
```
Comment (evykassirer) [inline test/evaluateFunctions.test.js]: can you add tests with substeps?

File: lib/evaluateFunctions.js [hunk 18]
```diff
@@ -96,4 +99,389 @@ function absoluteValue(node) {
...
+    }
+    else {
+      children.push(child);
+    }
+  }
+
+  if (children.length !== radicandNode.args.length) {
+    newNode.args[0] = NodeCreator.operator('*', children);
+    return NodeStatus.nodeChanged(
+      MathChangeTypes.FACTOR_INTO_PRIMES, node, newNode);
```
Comment (aelnaiem) [inline lib/evaluateFunctions.js]: Let's scrap it everywhere, as long as it doesn't change. In this case, the oldNode is always the original node since I separated each step into a different function. So as long as that's true, we should scrap it.

File: lib/evaluateFunctions.js [hunk 19]
```diff
@@ -96,4 +99,389 @@ function absoluteValue(node) {
...
+// factoring all the arguments under the radicand
+function factorChildren(node) {
+  let newNode = clone(node, false);
+  const radicandNode = node.args[0];
+
+  let children = [];
+  for (let i = 0; i < radicandNode.args.length; i++) {
+    const child = radicandNode.args[i];
+
+    if (NodeType.isConstant(child) && !Negative.isNegative(child)) {
```
Comment (aelnaiem) [inline lib/evaluateFunctions.js]: This wouldn't actually happen in the stepper as it would have been simplified, I'm only really testing for negative constants because we don't support complex numbers: sqrt(-1)

File: lib/evaluateFunctions.js [hunk 20]
```diff
@@ -96,4 +99,389 @@ function absoluteValue(node) {
...
+      children = children.concat(radicandNode.args.slice(i, j));
+    }
+    i = j;
+  }
+
+  // if we don't group any factors, then we can't simplify it any more
+  if (hasGroups) {
+    newNode.args[0] = children.length === 1 ?
+        children[0] : NodeCreator.operator('*', children);
+    return NodeStatus.nodeChanged(
```
Comment (aelnaiem) [inline lib/evaluateFunctions.js]: I see...

File: lib/evaluateFunctions.js (hunk 14 above)
Comment (aelnaiem) [inline lib/evaluateFunctions.js]: In this case specifically it's a multiplication node, because we've distributed the roots into the children

File: lib/evaluateFunctions.js [hunk 21]
```diff
@@ -96,4 +99,421 @@ function absoluteValue(node) {
...
+  }
+
+  return NodeStatus.noChange(node);
+}
+
+// Returns the nthRoot evaluated for a multiplication node.
+// Expects a multiplication node uder the radicand.
+// If the root is a positive constant, it factors the multiplicands, groups them
+// and converts the multiplications into exponents.
+// Distributes the nthRoot into the children nodes, evaluates those nthRoots
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: I think this could be easier to follow if the steps were numbered - find it a bit hard to follow still

File: lib/evaluateFunctions.js [hunk 22]
```diff
@@ -162,15 +165,19 @@ function nthRootExponent(node) {
...
-// Expects a multiplication node.
+// Expects a multiplication node uder the radicand.
+// If the root is a positive constant, it factors the multiplicands, groups them
+// and converts the multiplications into exponents.
+// Distributes the nthRoot into the children nodes, evaluates those nthRoots
+// and combines them
 function nthRootMultiplication(node) {
   let newNode = clone(node, false);
-  const rootNode = node.args.length === 2 ? node.args[1] : NodeCreator.constant(2);
+  const rootNode = getRootNode(node);
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: love it

File: lib/evaluateFunctions.js [hunk 23]
```diff
@@ -196,10 +203,8 @@ function nthRootMultiplication(node) {
   }
 
   status = distributeNthRoot(newNode);
-  if (status.hasChanged()) {
-    substeps.push(status);
-    newNode = NodeStatus.resetChangeGroups(status.newNode);
-  }
+  substeps.push(status);
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: what do you think about the `STEP X:` comments?

File: lib/evaluateFunctions.js (hunk 18 above)
Comment (evykassirer) [inline lib/evaluateFunctions.js]: cool, noted in the overall restructure list

File: lib/evaluateFunctions.js [hunk 24]
```diff
@@ -96,4 +99,389 @@ function absoluteValue(node) {
...
+    newNode.args[0] = NodeCreator.operator('*', children);
+    return NodeStatus.nodeChanged(
+      MathChangeTypes.FACTOR_INTO_PRIMES, node, newNode);
+  }
+
+  return NodeStatus.noChange(node);
+}
+
+// Given an nthRoot node with a constant positive root, will group the arguments
+// into groups of the root as a step
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: so it could group the arguments into groups of 2.3..?

and in general what happens for nthroot(x, y) where y isn't a whole number?

Comment (evykassirer) [inline lib/evaluateFunctions.js]: oo and what if it's 0? is a 0th root a thing?

File: lib/evaluateFunctions.js (hunk 20 above)
Comment (evykassirer) [inline lib/evaluateFunctions.js]: did you go through all your changes and see if they need change groups? or are you leaving that for your other task

File: lib/evaluateFunctions.js [hunk 25]
```diff
@@ -291,53 +289,50 @@ function groupTermsByRoot(node) {
...
       MathChangeTypes.GROUP_TERMS_BY_ROOT, node, newNode);
   }
+  // if we don't group any factors, then we can't simplify it any more
   return NodeStatus.noChange(node);
 }
 
-// Given an nthRoot node with a constant positive root, will group the arguments
+// Given an nthRoot node with a constant positive root,
 // will convert any grouped factors into exponent nodes as a step
 // e.g. nthRoot((2 * 2) * 2, 2) -> nthRoot(2^2 * 2, 2)
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: can you add a disclaimer that groups are expected to be the same length as the root number?

File: lib/evaluateFunctions.js [hunk 26]
```diff
@@ -382,26 +377,30 @@ function evaluateNthRootForChildren(node) {
...
+  }
+  else {
     return NodeStatus.nodeChanged(
       MathChangeTypes.EVALUATE_DISTRIBUTED_NTH_ROOT, node, newNode, true, substeps);
   }
-
-  return NodeStatus.noChange(node);
 }
 
 // Given a multiplication nodes, including nthRoots, will combine the nodes
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: given a root node with multiplication nodes? or just multiplication nodes. still not super clear cause the first thing you do is getRootNode on the node..

File: lib/evaluateFunctions.js [hunk 27]
```diff
@@ -382,26 +377,30 @@ function evaluateNthRootForChildren(node) {
...
 // with the same radicand as a step
 // e.g. 2 * nthRoot(2) * nthRoot(x) -> 2 * nthRoot(2 * x)
 function combineRoots(node) {
   let newNode = clone(node, false);
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: so here it seems that newNode is a root node

File: lib/evaluateFunctions.js [hunk 28]
```diff
@@ -477,11 +478,42 @@ function nthRootConstant(node) {
...
-// Simple helper function which determines if all the given nodes are equal
-function allEqual(nodes) {
-  return nodes.reduce((a, b) => {
+// Helpers
+
+// Given an nthRoot node, will return the root node.
+// The root node is the second child of the nthRoot node, but if one doesn't
+// exist, we assume it's a square root and return 2.
+function getRootNode(node) {
+  return node.args.length === 2 ? node.args[1] : NodeCreator.constant(2);
```
Comment (evykassirer) [inline lib/evaluateFunctions.js]: can you make it throw an error if it's not a root function node?

File: lib/evaluateFunctions.js (hunk 15 above)
Comment (aelnaiem) [inline lib/evaluateFunctions.js]: It doesn't, so I'm going to add that as future support

File: lib/evaluateFunctions.js (hunk 24 above)
Comment (aelnaiem) [inline lib/evaluateFunctions.js]: Seems like an edge case check, so I'm going to leave that as a future issue

File: lib/evaluateFunctions.js (hunk 27 above)
Comment (aelnaiem) [inline lib/evaluateFunctions.js]: Yea that's a mistake, it's not a root node. Fixing this now
